c2rust-analyze/src/main.rs

```diff
@@ -449,13 +449,117 @@ fn run(tcx: TyCtxt) {
+                    // For realloc and free, there is expected to be a cast to void*
+                    // from an arbitrary type in the same block.
+                    let mut handle_c_void_cast = |void_ptr: &Place<'tcx>,
+                                                  find_cast: &dyn Fn() -> Option<(
+                        Place<'tcx>,
+                        Place<'tcx>,
+                    )>| {
+                        let deref_ty = void_ptr
+                            .ty(acx.local_decls, acx.tcx())
+                            .ty
+                            .builtin_deref(true)
+                            .unwrap()
+                            .ty
+                            .kind();
+                        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+                            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+                            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
+                        });
+
+                        if let Some((casted, castee)) = find_cast() {
+                            if castee == *void_ptr {
+                                // This is a special case for types being casted from *c_void to a pointer
+                                // to some other type, e.g. `let foo = malloc(..) as *mut Foo;`
+                                // carry over the pointer id of *c_void, but match the pointer ids
+                                // of the casted-to-type for the rest
+                                acx.c_void_casts.0.insert(castee, casted);
+                            }
+                        }
+                    };
+
+                    let get_cast_op = |s: &Statement<'tcx>| match &s.kind {
+                        StatementKind::Assign(x) => match &x.1 {
+                            Rvalue::Cast(_, op, _) => Some((
+                                x.0,
+                                op.place().expect(
+                                    "Rvalue::Cast is not Operant::Constant so it has a Place",
+                                ),
+                            )),
+                            _ => None,
+                        },
+                        _ => None,
+                    };
+
+                    let find_last_cast_curr_block = || {
+                        bb_data
+                            .statements
+                            .iter()
+                            .rev()
+                            .find_map(|s| get_cast_op(s))
+                            .map(|(casted_to, casted_from)| (casted_from, casted_to))
+                    };
+
+                    let find_first_cast_succ_block = || {
+                        let successor_block_id = target.unwrap();
+                        mir.basic_blocks()[successor_block_id]
+                            .statements
+                            .iter()
+                            .find_map(|s| get_cast_op(s))
+                    };
```

Comment:
In the `malloc(...) as *mut T` pattern, can the `Rvalue::Cast` appear anywhere other than the first statement of the successor block?  My guess is that it can't (even in cases where the function arguments involve temporaries that need to be dropped, it looks like the drop happens after the cast).  If that's the case, then I think this should use something like `statements.first().filter_map(...)` rather than `statements.iter().find_map(...)` to ensure that it only considers the first statement in the block.

Reply:
> In the malloc(...) as *mut T pattern, can the Rvalue::Cast appear anywhere other than the first statement of the successor block?

Yes it can. We can have `StorageDead` statements in between. Should those be ignored? If they are ignored then I think you may be correct in that the cast is the next type of statement.

Reply:
>We can have StorageDead statements in between. Should those be ignored?

Yes, I think we should just ignore `StorageDead` at the start of the successor block.

`StorageDead` corresponds to `@llvm.lifetime.end`, which overwrites all bytes of the value with poison.  If we were doing an actual code transformation on the MIR, we'd need to be careful about where exactly we place those relative to the call, and we'd probably need to check for overlap between `StorageDead` locals and locals that appear in the inputs/outputs of the call and the cast.  But for the flow-insensitive analysis that we're actually using, that ordering of statements doesn't matter.

Reply:
@spernsteiner I have something in progress in #848, would you mind taking a look? I'm almost positive that this won't be the case in the reverse with `realloc`, however. There can be statements and entire blocks between casts to `void*` and the `realloc` call. Because there won't be a fixed distance between the two in this case, and no longer a fixed distance between allocations and casts from void, I added something in #848 to key on the location of both the call and the cast. The former to know what updated `Place` to use, and the latter to know whether or not to skip a cast in `visit_statement`